internal/registry/json.go

```diff
@@ -0,0 +1,151 @@
+	History      []historyV1 `json:"history,omitempty"`
+}
+
+type historyV1 struct {
+	CreatedBy  string `json:"created_by"`
+	EmptyLayer bool   `json:"empty_layer,omitempty"`
+}
+
+// imageIndexV1 represents OCI Registry "/v2/${Repository}/manifests/${Tag}" responses for these media-types:
+// * "application/vnd.oci.image.index.v1+json"
+// * "application/vnd.docker.distribution.manifest.list.v2+json"
+//
+// See https://github.com/opencontainers/image-spec/blob/master/schema/image-index-schema.json
+type imageIndexV1 struct {
+	Manifests []imageManifestReferenceV1 `json:"manifests"`
+}
+
+type imageManifestReferenceV1 struct {
+	MediaType string     `json:"mediaType"`
+	Digest    string     `json:"digest"`
+	Platform  platformV1 `json:"platform"`
+}
+type platformV1 struct { // redefined here because of the dotted "os.version" json field name.
+	Architecture string `json:"architecture"`
+	OS           string `json:"os"`
+	OSVersion    string `json:"os.version,omitempty"`
+}
+
+// imageManifestV1 represents responses matching isMediaTypeImageManifestV1
+
+// imageManifestV1 represents OCI Registry "/v2/${Repository}/manifests/${Tag}" responses for these media-types:
+// * "application/vnd.oci.image.manifest.v1+json"
+// * "application/vnd.docker.distribution.manifest.v2+json"
+//
+// See https://github.com/opencontainers/image-spec/blob/master/schema/image-manifest-schema.json
+type imageManifestV1 struct {
+	URL    string         // not in the JSON
+	Config descriptorV1   `json:"config"`
+	Layers []descriptorV1 `json:"layers"`
+}
+
+// See https://github.com/opencontainers/image-spec/blob/master/descriptor.md
+type descriptorV1 struct {
+	MediaType string `json:"mediaType"`
+	Digest    string `json:"digest"`
+	Size      int64  `json:"size"`
+}
+
+var (
+	// ignoredDockerDirectives are Dockerfile directives that don't result in a tarball which could contain a binary.
+	// This is used because some versions of Docker don't set `"empty_layer": true` in the config JSON.
+	// We can't use an allow list because "RUN", "ADD" and "COPY" are not always in "created_by", most notably in the
+	// canonical images made by https://github.com/docker-library/bashbrew
+	ignoredDockerDirectives = []string{
+		"ARG",
+		"CMD",
+		"ENTRYPOINT",
+		"ENV",
+		"EXPOSE",
+		"HEALTHCHECK",
+		"LABEL",
+		"MAINTAINER",
+		"ONBUILD",
+		"SHELL",
+		"STOPSIGNAL",
+		"USER",
+		"VOLUME",
+		"WORKDIR",
+	}
+	// skipCreatedByPattern is currently Windows-specific, but might be needed on other images if we find cases of it.
+	// There are intentionally multiple spaces allowed as there are examples of it, caused by Moby joining on space:
+	// https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/image/v1/imagev1.go#L32
+	skipCreatedByPattern = regexp.MustCompile(".* +(?:" + strings.Join(ignoredDockerDirectives, "|") + ") .*")
+)
+
+func newImage(baseURL string, i *imageManifestV1, c *imageConfigV1) *internal.Image {
```

Comment:
this is the important code. it uses all the stored knowledge in comments and test jsons to make `Image` which is actually usable.

Reply:
shall we use the explanatory name for args? hard to understand what `i` means  with `j, k` being used below.

Reply:
I would suggest 
```suggestion
func newImage(baseURL string, manifest *imageManifestV1, config *imageConfigV1) *internal.Image {
```